zrml/hybrid-router/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+                        // existing order is willing to give the required `asset` as the `maker_asset`
+                        ensure!(
+                            asset == order.maker_asset,
+                            Error::<T>::AssetNotEqualToOrderBookMakerAsset
+                        );
+                        ensure!(order_price <= price_limit, Error::<T>::OrderPriceAboveMaxPrice);
+                    }
+                    TxType::Sell => {
+                        // existing order is willing to receive the required `asset` as the `taker_asset`
+                        ensure!(
+                            asset == order.taker_asset,
+                            Error::<T>::AssetNotEqualToOrderBookTakerAsset
+                        );
+                        ensure!(order_price >= price_limit, Error::<T>::OrderPriceBelowMinPrice);
+                    }
+                }
+
+                let amm_trade_info = Self::maybe_fill_from_amm(
+                    tx_type,
+                    who,
+                    market_id,
+                    asset,
+                    remaining,
+                    order_price,
+                )?;
+
+                if let Some(t) = amm_trade_info.1 {
+                    amm_trades.push(t);
+                }
+                remaining = amm_trade_info.0;
+
+                if remaining.is_zero() {
+                    break;
+                }
+
+                // `remaining` is always denominated in the `taker_asset`
+                // because this is what the order owner (maker) wants to receive
+                let (_taker_fill, maker_fill) =
+                    order.taker_and_maker_fill_from_taker_amount(remaining)?;
+                // and the `maker_partial_fill` of `fill_order` is specified in `taker_asset`
+                let order_trade_opt = Self::handle_fill_order(who.clone(), order_id, maker_fill)?;
+                if let Some(order_trade) = order_trade_opt {
+                    order_trades.push(order_trade);
+                    // `maker_fill` is the amount the order owner (maker) wants to receive
+                    remaining = remaining.checked_sub_res(&maker_fill)?;
+                }
+            }
+
+            Ok(OrderAmmTradesInfo { remaining, order_trades, amm_trades })
+        }
+
+        fn handle_fill_order(
+            who: AccountIdOf<T>,
+            order_id: OrderId,
+            maker_fill: BalanceOf<T>,
+        ) -> Result<Option<OrderTradeOf<T>>, DispatchError> {
+            match T::OrderBook::fill_order(who, order_id, Some(maker_fill)) {
+                Ok(order_trade) => Ok(Some(order_trade)),
+                Err(ApiError::SoftFailure(OrderbookSoftFail::BelowMinimumBalance))
+                | Err(ApiError::SoftFailure(
+                    OrderbookSoftFail::PartialFillNearFullFillNotAllowed,
+                )) => Ok(None),
+                Err(ApiError::HardFailure(dispatch_error)) => Err(dispatch_error),
```

Comment:
Both of these paths look like they're untested.

Reply:
7fb7705781e1efbe08f8398e4df0d9158e8392cc